### ISSUE TITLE
ArkMapPlot support for ARK-Tools v0.6.0

### ISSUE THREAD

**Issue by mitfry** [shadowmage45/ArkMapPlot#2]
Yo ShadowMage,

As I updated your tool I also found a new version of the ARK-Tools. Some changes have been made to the way JSON files are looking, which seems to cause the tool to crash.

For now I'm just using an older version of ARK-tools to generate the JSON files for ArkMapPlot to read.

Would you have time to take a look at this?

**Comment by shadowmage45** [comment]
If you could post up an example of one of the new JSON files, I can at least take a look at it (certainly don't have time to re-install ARK, create a new game, download/install ark-tools, do a data-dump, etc..).

No guarantees though; it could be a simple syntax/formatting change (simple/quick to fix), or it could be the entire layout and structure that has changed (harder to fix).

**Comment by mitfry** [comment]
I'll get right on it. Give me a couple of minutes.

**Comment by mitfry** [comment]
I've included the output JSON file of Rexes from an Island map datafile. Because I couldn't upload a .json file I changed it to a .txt file. Just remove".txt" at the end of the file to get the original file back.

This is the changelog from v0.6.0. Maybe there is some useful information in there:
Changelog:

- Replaced javax.json with Jackson.
- Moved JsonDriver property writers to properties owning class.
- Fixed JsonDriver never finishing.
- Adjustments to the names of some fields.
- Use property writers to replace old GameObject to JSON code.
- Allow "--write-all-fields" for most commands.
- "--allow-broken-file" is now available for all xToJson commands.
- Allow parallel processing for players, tribes and cluster commands.
- Amount of parallelism can no be controlled via --thread-count, defaults to the number of cores available to the JVM.
- All async options are gone.
- Updated to newest version of ark-savegame-toolkit, adds full support for Version 7 to 9 map save formats, Version 4 LocalProfile and Version 4 CloudInventory.
- New parameters for player, tribes and creatures commands.

[Rex_Character_BP_C.json.txt](https://github.com/shadowmage45/ArkMapPlot/files/1303609/Rex_Character_BP_C.json.txt)

Is there anything else I need to provide?

**Comment by shadowmage45** [comment]
Hopefully that json file should be enough.  Will try to get some testing done sometime soon-ish, but can't guarantee anything.

A quick looks tells me that the location (x,y,z,lat,lon) data was moved into a 'location' sub-node, rather than being stored in the root node.  This alone could cause the crashing, and if that is the only problem, should be quick enough to fix up.

**Comment by shadowmage45** [comment]
Old format:
````
    {
        "x":-227529.328125,
        "y":148138.765625,
        "z":-4955.85302734375,
        "lat":68.5,
        "lon":21.6,
        "id":1475450497676762242,
        "color0":27,
        "color1":34,
        "color3":35,
        "color5":33,
        "baseLevel":96,
        "wildLevels":
        {
            "health":14,
            "stamina":19,
            "oxygen":15,
            "food":15,
            "weight":13,
            "melee":6,
            "speed":13
        }
    },
````
New format:
````
     {
        "baseLevel" : 75,
        "colorSetIndices" : 
        {
            "0" : 22,
            "1" : 36,
            "3" : 22,
            "5" : 22
        },
        "female" : true,
        "id" : 1266527461647920456,
        "lastEnterStasisTime" : 6.39707883411075E7,
        "location" :
        {
            "x" : 88175.43,
            "y" : -75131.29,
            "z" : 2299.3381,
            "lat" : 61.021927,
            "lon" : 40.60859
        },
        "requiredTameAffinity" : 14700.0,
        "tamingEffectivness" : 1.0,
        "team" : 7,
        "type" : "Rex",
        "wildLevels" :
        {
            "health" : 12,
            "stamina" : 11,
            "oxygen" : 9,
            "food" : 12,
            "weight" : 8,
            "melee" : 10,
            "speed" : 12
        }
    },
````

**Comment by mitfry** [comment]
Maybe this image will help, Left is old and right is the new format.

![image](https://user-images.githubusercontent.com/526358/30441579-25026a8a-997a-11e7-88e5-24b9e358d6af.png)

**Comment by mitfry** [comment]
Oh, you just posted the same thing, lol.

It seems the Colors and the Location info is now formatted like the Wildlevels were in the old format. There also seem to be more fields. Correct?

**Comment by shadowmage45** [comment]
Yeah, that seems to be the majority of the changes that would effect the reading/loading of those files.

Now that I know a bit about what I'm looking for, it shouldn't take too long to find/fix the code to read the new format (but it will lose the ability to read the old format at the same time).

**Comment by mitfry** [comment]
Alright, great! Maybe you could make a release of the current version that works for the old format and clearly note that it ONLY supports ARK-Tools exports until v0.5.4.

Going forward people would just have to download the latest version of your code that works for v0.6.0 and up. 

I found the following new data fields:
- colorSetIndices (sub node)
- lastEnterStatisTime
- location (sub node)
- requiredTameAffinity
- tamingEffectivness (which should probably be "tamingEffectiveness")
- type

Would it be an idea to have the data sorted from high to low automatically? Sorting by the lvl column for example.

**Comment by mitfry** [comment]
A similar change to the JSON files might happen again in the near future:

UPCOMING:
0.6.1
- Fix for missing ancestor writer, include names of ancestors

0.7.0
- Breaking change of xToJson formats (again), improved support for native array properties (properties which may or may not be ArrayProperty but still contain multiple values)
- Fix for objects in the hibernation section being ignored by all commands
- Implement missing stuff to make the database system useful (e.G. json driver should be able to write data to multiple files, skip refreshing files without changes, implement missing support for cluster directory)
- Add more things to the modify command

**Comment by shadowmage45** [comment]
Just uploaded an updated version that -should- fix the problems regarding loading the new .json format.  -Should-.  Completely untested so far (need to get home before I can do much testing; currently doing this all from work).

**Comment by mitfry** [comment]
Thank you! I've downloaded the latest source and couldn't compile. Getting an error. Not sure if it'll help you, but this is the error log:

[CompileError.txt](https://github.com/shadowmage45/ArkMapPlot/files/1304560/CompileError.txt)

It did however generate a new ArkMapPlot.exe but it wouldn't start with the new JSON files. 

Let me know if I can or need to provide anything else.

**Comment by shadowmage45** [comment]
Hmm.. thanks for taking a look at it.  Not much else you can do -- I'm going to have to compile it myself and run some tests with the example .json file you sent (will probably have to put additional debugging in place to find where the errors are at, as debugging c# WPF apps is a bit of an arcane art).

**Comment by shadowmage45** [comment]
Turns out it was working properly (the code), but was crashing on stale data left in the debug/config folders.

Try using the setup from this .zip:
[Release.zip](https://github.com/shadowmage45/ArkMapPlot/files/1304983/Release.zip)

**Comment by mitfry** [comment]
I can confirm the release from your .zip to be working with ARK-tools v0.6.0. 

Thanks a lot for your time (also while at work, hehe)! I'll let you know when ARK-Tools makes significant changes to the JSON files again.
